### PR TITLE
Prepare API for upcoming UI changes

### PR DESCRIPTION
**Pull request by ganhongyao**
Changes
- Add primary donor logo (frontend + backend)
- Add campaign image in interest (frontend + backend)
- Replace `is_redeemed` column in `Coupons` table with `campaign_charity_id`
   - This is because the new schema is such that if a coupon is redeemed without any personal contributions, there are no secondary donations. We need `campaign_charity_id` to keep track which campaign charity this coupon was redeemed for.
   - This makes `redeemed?` derived from whether the `campaign_charity_id` is null or not
- Rework `secondary_donations` POST endpoint

| Behaviour                                        | Endpoint                 | Request body type |
|--------------------------------------------------|--------------------------|--------------|
| Redeeming a coupon with personal contribution    | POST coupons/redeem      |    CouponRedeemPostData         |
| Redeeming a coupon without personal contribution | POST coupons/redeem      |   CouponRedeemPostData, with amount = 0        |
| Making an external contribution without coupon   | POST secondary_donations |     SecondaryDonationPostData         |
- Update views as necessary for upcoming UI changes

### CODE REVIEW

File: backend/app/controllers/secondary_donations_controller.rb [hunk 1]
```diff
@@ -9,7 +9,6 @@ def index
 
   def create
     @secondary_donation = SecondaryDonation.new(secondary_donation_params)
-    @secondary_donation.coupon = Coupon.find_by(url_token: params[:url_token])
     @secondary_donation.campaign_charity = CampaignCharity.find(params[:campaign_charity_id])
     @secondary_donation.save!
```
Comment (ganhongyao) [inline backend/app/controllers/secondary_donations_controller.rb:13]: This endpoint is now only used for external donations (without coupons)

**Comment by sivayogasubramanian** [comment]
I am going to request changes as I come across them when I am doing the frontend components.

frontend/components/campaigns/campaignCharities/CampaignCharityCard.tsx in #252 would need PrimaryDonor name in the CampaignCharityDonationPublicData.

File: frontend/types/campaignCharities.ts [hunk 2]
```diff
@@ -14,8 +15,8 @@ export type CampaignCharityData = CampaignCharityBaseData & {
   charity: CharityData;
 };
 
-export type CampaignCharityDonationPublicData = CampaignCharityData & DonationBreakdownData;
+export type CampaignCharityDonationPublicData = CampaignCharityData & DonationBreakdownData & PrimaryDonorData;
 
-export type CampaignCharityDonationData = CampaignCharityData & DonationBreakdownData;
+export type CampaignCharityDonationData = CampaignCharityData & DonationBreakdownData & PrimaryDonorData;
```
Comment (sivayogasubramanian) [inline frontend/types/campaignCharities.ts:20]: Doesn't some fields clash. ID in CampaignCharityData and PrimaryDonorData?

File: backend/app/views/campaign_charities/_donation_public.json.jbuilder [hunk 3]
```diff
@@ -2,5 +2,7 @@
 
 json.partial! 'campaign_charities/campaign_charity', campaign_charity: campaign_charity
 json.partial! 'secondary_donations/breakdown', donation_breakdown: campaign_charity.donation_breakdown
-
+json.primaryDonor do
+  json.partial! 'primary_donors/primary_donor', primary_donor: campaign_charity.campaign.primary_donor
+end
```
Comment (sivayogasubramanian) [inline backend/app/views/campaign_charities/_donation_public.json.jbuilder:7]: This primaryDonor key is not in the type

**Comment by sivayogasubramanian** [comment]
I made the fixes in #252. Help check if its okay.

**Comment by zognin** [comment]
I didn't realise siva's PRs base was this branch, so I reverted everything

File: frontend/types/campaignCharities.ts (hunk 2 above)
Comment (ganhongyao) [reply frontend/types/campaignCharities.ts]: Meant for PrimaryDonorData to be in a nested object `primaryDonor`, fixed

**Review by sivayogasubramanian — APPROVED** [review]
LGTM! Didn't really test everything as it will be easier to find issues during integration.

**Pull request merged**